### PR TITLE
bt-516: Connect search to API

### LINKED ISSUE #516: Connect search to API
**Acceptance criteria:**

- Employer should be able to search for talents
- All talents who created profiles and were verified by admin should be displayed on the search page
- Employer should be able to search using filters: Job Title, Hard Skills, Years of experience, BSA Characteristics, BSA Badges, BSA Project, Location, Level of English, Employment Type.
- Employer should be able to search using: basic search and full-text search 
- All talents by the selected request should be displayed

### LINKED ISSUE #539: Remove "BSA badges" filter from Search page
Image

### PR DESCRIPTION
**Pull request by nika-sirak**
This PR is related to #516 and #539 tickets.

https://github.com/BinaryStudioAcademy/bsa-2023-bsa-talents/assets/60103912/ba5b25df-a03a-469a-9d6c-fdadcdd89497

### CODE REVIEW

File: shared/src/helpers/configure-string/configure-string.helper.ts
Comment (nikita-remeslov) [inline shared/src/helpers/configure-string/configure-string.helper.ts:1]: why so complicated?

File: frontend/src/bundles/employers/pages/candidates.tsx [hunk 1]
```diff
@@ -53,10 +51,14 @@ const Candidates: React.FC = () => {
         useAppForm<EmployeesFiltersDto>({
             defaultValues: DEFAULT_EMPLOYEES_FILTERS_PAYLOAD,
```
Comment (savinaDaria) [inline frontend/src/bundles/employers/pages/candidates.tsx]: when navigating to another page and then return - filters are reset, I mean if I go to the candidate page and then back to search I have all filters default
try to use [this changes](https://github.com/BinaryStudioAcademy/bsa-2023-bsa-talents/pull/625/commits/687988e906e076eafa66a1aee0c01f49119d65e5) from draft pr 
here they are not reset but then got a problem with clearing filters to the default with Clear filters button

Comment (savinaDaria) [reply frontend/src/bundles/employers/pages/candidates.tsx]: actually fixed _Clear filters handling_ in [this commit](https://github.com/BinaryStudioAcademy/bsa-2023-bsa-talents/pull/625/commits/0b5cd42d558c01c61f1cce4d09ee2c16d1565d46)
So you can add changes from these 2 commits to your PR

File: shared/src/helpers/configure-string/configure-string.helper.ts
Comment (nika-sirak) [reply shared/src/helpers/configure-string/configure-string.helper.ts:1]: Updated

File: frontend/src/bundles/employers/pages/candidates.tsx (hunk 1 above)
Comment (nika-sirak) [reply frontend/src/bundles/employers/pages/candidates.tsx]: Added

File: shared/src/helpers/configure-string/map-query-object-to-string.ts [hunk 2]
```diff
@@ -0,0 +1,24 @@
...
+        })
+        .flatMap(([key, value]) => {
+            return Array.isArray(value)
+                ? value.map((v) => `${key}=${v}`)
+                : `${key}=${value}`;
+        })
+        .join('&');
+};
+
+export { mapQueryObjectToString };
```
Comment (nikita-remeslov) [inline shared/src/helpers/configure-string/map-query-object-to-string.ts]:
```suggestion
const mapQueryObjectToString = <T extends Record<string, string | string[] | boolean | number | number[]>>(
  query: T
): string => {
  return Object.entries(query)
    .filter(([_, value]) => value !== '' && !(Array.isArray(value) && value.length === 0))
    .flatMap(([key, value]) =>
      Array.isArray(value) ? value.map((v) => `${key}=${v}`) : `${key}=${value}`
    )
    .join('&');
};

export { mapQueryObjectToString };
```

**Review by nikita-remeslov — CHANGES_REQUESTED** [review]

File: backend/src/bundles/user-details/user-details.repository.ts [hunk 3]
```diff
@@ -178,11 +170,16 @@ class UserDetailsRepository implements Repository {
             sortingParameters.direction,
         );
 
-        const searchResults = await query;
+        const searchResults = await query
+            .withGraphJoined('user')
+            .where('user.role', '=', 'talent');
```
Comment (MaksymGavatuk) [inline backend/src/bundles/user-details/user-details.repository.ts]: We have enum for 'talent' ?

File: frontend/src/bundles/talent-onboarding/components/candidate-profile/profile-first-section/profile-first-section.tsx [hunk 4]
```diff
@@ -60,8 +60,8 @@ const ProfileFirstSection: React.FC<Properties> = ({
                     >
                         {candidateParameters.location} |{' '}
                         {candidateParameters.experienceYears} years of
-                        experience |{candidateParameters.englishLevel} | Publish
-                        today
+                        experience |{candidateParameters.englishLevel} |
+                        Published {candidateParameters.date.split('T')[0]}
```
Comment (MaksymGavatuk) [inline frontend/src/bundles/talent-onboarding/components/candidate-profile/profile-first-section/profile-first-section.tsx]: Can we use date formatting for this ?

**Review by MaksymGavatuk — CHANGES_REQUESTED** [review]

File: frontend/src/bundles/talent-onboarding/components/candidate-profile/profile-first-section/profile-first-section.tsx (hunk 4 above)
Comment (nika-sirak) [reply frontend/src/bundles/talent-onboarding/components/candidate-profile/profile-first-section/profile-first-section.tsx]: Fixed

**Comment by vladjsp** [comment]
I think we need some simple informational text when the search result is empty because now the query is done but the search results is empty and it can be confusing.

**Comment by nika-sirak** [comment]
We have radio button 'Basic Search' and isActiveSearch checkbox for default search. 

image

**Review by nikita-remeslov — APPROVED** [review]

File: backend/src/bundles/user-details/user-details.controller.ts [hunk 5]
```diff
@@ -209,17 +209,18 @@ class UserDetailsController extends ControllerBase {
         });
 
         this.addRoute({
-            path: UserDetailsApiPath.ROOT,
+            path: '',
```
Comment (nikita-remeslov) [inline backend/src/bundles/user-details/user-details.controller.ts]: i think we dont need this change!

**Review by nikita-remeslov — CHANGES_REQUESTED** [review]

**Review by vladjsp — APPROVED** [review]
Works well. Tested locally.

File: backend/src/bundles/user-details/user-details.controller.ts (hunk 5 above)
Comment (nika-sirak) [reply backend/src/bundles/user-details/user-details.controller.ts]: Fixed

**Review by nikita-remeslov — APPROVED** [review]

**Pull request merged**